plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs

```diff
@@ -292,45 +397,43 @@ mkDatatypeValueType ann (Datatype _ tn tvs _ _)  = PIR.mkIterTyApp ann (PIR.mkTy
+type variables free.
+@
+    mkConstructorType List Cons = forall (a :: *) . a -> List a -> List a
+@
+-}
 mkConstructorType :: Datatype TyName Name uni (Provenance a) -> VarDecl TyName Name uni (Provenance a) -> PIRType uni a
 -- this type appears *inside* the scope of the abstraction for the datatype so we can just reference the name and
 -- we don't need to do anything to the declared type
 -- see note [Abstract data types]
 -- FIXME: normalize constructors also here
-mkConstructorType (Datatype _ _ tvs _ _) constr =
-    let constrTy = PIR.mkIterTyForall tvs $ _varDeclType constr
-    in fmap (\a -> DatatypeComponent ConstructorType a) constrTy
-
--- See note [Scott encoding of datatypes]
--- | Make a constructor of a 'Datatype' with the given pattern functor. The constructor argument mostly serves to identify the constructor
--- that we are interested in in the list of constructors.
--- @
---     mkConstructor <definition of List> <pattern functor of List> List Cons
---         = /\(a :: *) . \(arg1 : a) (arg2 : <definition of List> a) .
---             wrap <pattern functor of List> /\(out_List :: *) .
---                 \(case_Nil : out_List) (case_Cons : a -> List a -> out_List) . case_Cons arg1 arg2
--- @
-mkConstructor :: MonadQuote m => PLCRecType uni fun a -> Datatype TyName Name uni (Provenance a) -> Int -> m (PIRTerm uni fun a)
-mkConstructor dty d@(Datatype ann _ tvs _ constrs) index = do
-    resultType <- resultTypeName d
-
-    -- case arguments and their types
-    casesAndTypes <- do
-          -- these types appear *outside* the scope of the abstraction for the datatype, so we need to use the concrete datatype here
-          -- see note [Abstract data types]
-          -- FIXME: normalize datacons' types also here
-          let caseTypes = unveilDatatype (getType dty) d <$> fmap (constructorCaseType (TyVar ann resultType)) constrs
-          caseArgNames <- for constrs (\c -> safeFreshName $ "case_" <> T.pack (varDeclNameString c))
-          pure $ zipWith (VarDecl ann) caseArgNames caseTypes
-
+mkConstructorType (Datatype _ _ tvs _ _) constr = PIR.mkIterTyForall tvs $ _varDeclType constr
+
+-- See note [Encoding of datatypes]
+{- | Make a constructor of a 'Datatype' with the given pattern functor. The constructor argument mostly serves to identify the constructor
+that we are interested in in the list of constructors.
+
+Scott:
+@
+    mkConstructor <definition of List> List Cons
+        = /\(a :: *) . \(arg1 : a) (arg2 : <definition of List> a) .
+            wrap <pattern functor of List> /\(out_List :: *) .
+                \(case_Nil : out_List) (case_Cons : a -> List a -> out_List) . case_Cons arg1 arg2
+@
+
+SOPs:
+@
+    mkConstructor <definition of List> List Cons
+        = /\(a :: *) . \(arg1 : a) (arg2 : <definition of List> a) .
+            wrap <pattern functor of List> /\(out_List :: *) .
+                constr ((\(List :: * -> *) . <pattern functor of List>) <definition of List>) arg1 arg2
```

Comment:
Those two occurrences of `<pattern functor of List>` make it seem like they are the same thing, but they're not:

1. the latter is a pattern functor body, not an entire pattern functor
2. the former is what I call "packed" in the compilation machinery, i.e. 1-ary one after all the encoding tricks are done, and the latter is the body of an original "unpacked" n-ary pattern functor

Reply:
I'm leaving this in because I don't understand it :/